go/vt/vtgate/evalengine/compiler_json.go

```diff
@@ -66,14 +66,34 @@ func (c *compiler) compileToJSON(doct ctype, offset int) (ctype, error) {
 	return ctype{Type: sqltypes.TypeJSON, Col: collationJSON}, nil
 }
 
+func (c *compiler) compileArgToJSON(doct ctype, offset int) (ctype, error) {
```

Comment:
We need separate handling for an argument to say `JSON_OBJECT` to a partial JSON object. For example, a JSON string on its own is not a valid JSON object but for `JSON_OBJECT` we need to allow partial JSON types so we can instantiate the right one.

On the other hand, usages like `CAST('' as json)` evaluate if the string is a full JSON object and don't allow partial things like strings so it's different. So hence a new function here.